internal/monitors/postgresql/monitor.go

```diff
@@ -0,0 +1,213 @@
+package postgresql
+
+import (
+	"context"
+	dbsql "database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/signalfx/signalfx-agent/internal/core/config"
+	"github.com/signalfx/signalfx-agent/internal/monitors"
+	"github.com/signalfx/signalfx-agent/internal/monitors/sql"
+	"github.com/signalfx/signalfx-agent/internal/monitors/types"
+	"github.com/signalfx/signalfx-agent/internal/utils"
+	"github.com/signalfx/signalfx-agent/internal/utils/filter"
+	"github.com/sirupsen/logrus"
+)
+
+const monitorType = "postgresql"
+
+var logger = logrus.WithFields(logrus.Fields{"monitorType": monitorType})
+
+func init() {
+	monitors.Register(monitorType, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// Config for the postgresql monitor
+type Config struct {
+	config.MonitorConfig `yaml:",inline" acceptsEndpoints:"true"`
+
+	Host string `yaml:"host"`
+	Port uint16 `yaml:"port"`
+
+	// See
+	// https://godoc.org/github.com/lib/pq#hdr-Connection_String_Parameters.
+	ConnectionString string `yaml:"connectionString"`
+
+	// List of databases to send database-specific metrics about.  If omitted, metrics about all databases will be sent.  This is an [overridable filter](https://github.com/signalfx/signalfx-agent/blob/master/docs/filtering.md#overridable-filters).
+	Databases []string `yaml:"databases"`
+
+	// How frequently to poll for new/deleted databases in the DB server.
+	// Defaults to the same as `intervalSeconds` if not set.
+	DatabasePollIntervalSeconds int `yaml:"databasePollIntervalSeconds"`
+}
+
+func (c *Config) connStr() string {
+	connStr := c.ConnectionString
+	if c.Host != "" {
+		connStr += " host=" + c.Host
+	}
+	if c.Port != 0 {
+		connStr += fmt.Sprintf(" port=%d", c.Port)
+	}
+
+	return connStr
+}
+
+// Monitor that collects postgresql stats
+type Monitor struct {
+	sync.Mutex
+
+	Output types.Output
+	ctx    context.Context
+	cancel context.CancelFunc
+	conf   *Config
+
+	database *dbsql.DB
+
+	monitoredDBs  map[string]*sql.Monitor
+	serverMonitor *sql.Monitor
+}
+
+// Configure the monitor and kick off metric collection
+func (m *Monitor) Configure(conf *Config) error {
+	m.conf = conf
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	var err error
+	m.database, err = dbsql.Open("postgres", conf.connStr()+" dbname=postgres")
+	if err != nil {
+		return err
+	}
+
+	var dbFilter filter.StringFilter
+	if len(conf.Databases) > 0 {
+		dbFilter, err = filter.NewExhaustiveStringFilter(conf.Databases)
+		if err != nil {
+			return fmt.Errorf("problem with databases filter: %v", err)
```

Comment:
Does `Shutdown()` get called to cleanup stuff if `Configure()` fails?

Reply:
No, good point.  The Configure method should be responsible for closing the DB in the error case. Will update.